src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +409,115 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
             }
         }
 
+        private async Task<(CompletionList CompletionList, bool IsIncomplete, long ResultId)?> GetCompletionListAndFilterDownAsync(
+            LSP.CompletionParams request,
+            SourceText sourceText,
+            Document document,
+            OptionSet completionOptions,
+            CompletionService completionService,
+            CancellationToken cancellationToken)
+        {
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+            var completionTrigger = await ProtocolConversions.LSPToRoslynCompletionTriggerAsync(request.Context, document, position, cancellationToken).ConfigureAwait(false);
+
+            // Get the cached completion list in the case of a request for an incomplete list or calculate a new list and cache it.
+            CompletionList completionList;
+            long resultId;
+            if (request.Context?.TriggerKind == LSP.CompletionTriggerKind.TriggerForIncompleteCompletions)
+            {
+                // This is a request for a list we already calculated but was incomplete.
+                // Use the resultId to retrieve the completion list we cached for the original result.
+                Contract.ThrowIfFalse(_lastIncompleteResultId.HasValue, "Trigger for incomplete list missing previous result id");
+                var cachedList = _completionListCache.GetCachedCompletionList(_lastIncompleteResultId.Value);
+                Contract.ThrowIfNull(cachedList, "Trigger for incomplete list missing cached list");
```

Comment:
Presumably the platform requeries us for the list when there is an error? We don't need to return the list on a cache miss?

Reply:
this check feels weird.  i'm ok with us be stateful as an optimization.  but say we got recycled but completion still calls to us with this value.  Either:

1. this is ok, and we shoudl be resilient and should respond with some sensible message.
2. this is not ok, and we shoudl have caught this when we first got he message and we were validating it for invariant correctness.

`Contract` calls are for catching when we make mistakes entirely in *our* domain.  in other words, this shouldn't be possible given everythign we know about the code under *our* control.

Reply:
I can move the contract checks up.  Being resilient here is difficult - the completion service only returns results for the first character typed on the word.  So if we get an incomplete completion request in the middle of a word, even if we asked the completion service for more items, we wouldn't get any items.

Perhaps one solution would be to figure out the filter text, and just re-request completions on a document containing only the word start character?  I'll try something like that out

Reply:
Updated to re-calculate the list using a document w/out the filter text